content/blog/2019-10-23-register-allocation/index.md

```diff
@@ -0,0 +1,142 @@
+Only the variables with a assigned color will be allocated to the corresponding registers, all reads or writes to other variables will involve a memory load and store.
+
+### Code Generation
+Once we have the graph coloring scheme, we start to generate a new Bril program with register allocation.
+We use special variables `r_xx` to represent registers.  For example, 
+```
+a: int = const 5;
+b: int = const 3;
+c: int = add a b;
+print c;
+```
+with `r_01` allocated to `a` and `r_02` allocated to `c` will look like the following:
+```
+r_01: int = const 5;
+b: int = const 3;
+r_02: int = add r_01 b;
+print r_02;
+```
+
+If any data transfering between registers and stack is needed, it could be done by `id` operations like `a: int = id r_01`.
+
+## Evaluation Method
+We evaluate our implementation by counting load and store counts during program execution.
+
+### Benchmarks
+We wrote some common kernels and some hand written programs that can be mapped with a specific number of registers to test our register allocation performance. Following table provides a list of these benchmarks.
+
+| Test   | Description |
+|--------|---------------------------------------------------------------------------------------|
+| br     | Tests with branching                                                          |
+| if     | Tests with a more complex conditional                                        |
+| loop   | Tests with a loop                                                          |
+| alloc1 | Tests a simple 4 variable program taken from [these](https://web.stanford.edu/class/archive/cs/cs143/cs143.1128/lectures/17/Slides17.pdf) slides.    |
+| alloc2 | Test with 14 variables possible to allocate to 4 registers, inspired from the [book](https://www.google.com/books/edition/Modern_Compiler_Implementation_in_Java/N-sgAwAAQBAJ?hl=en&gbpv=0) |
+| alloc3 | Test with 14 variables possible to allocate to 6 registers, adopted from the [book](https://www.google.com/books/edition/Modern_Compiler_Implementation_in_Java/N-sgAwAAQBAJ?hl=en&gbpv=0) |
+| matmul | An implementation of matrix multiplication in Bril                                |
+| dotprod | An implementation of dot product in Bril                                |
+| fib | An implementation of fibonnacci in Bril                                |
+| fact | An implementation of factorial in Bril                                |
+| polymul | An implementation of polynomial multiplication in Bril                    |
+
+The last 5 benchmarks are from this [repository](https://github.com/xu3kev/bril/tree/master/benchmark).
+
+### Baseline
+In order to evaluate our register allocation, we assume that there is no register holding data across different instruction as the baseline, so that each instruction loads all operands from the memory and stores the result back to memory after execution. This is a conservative approach and provides an upper bound to the number of loads and stores that can occur.
```

Comment:
> we assume that there is no register holding data across different instruction as the baseline

This is a little hard to parse—are you saying that, in your assumed machine model, operations can happen "directly" on memory operands without first going into a register? FWIW, there is a term "register machine" for an ISA where everything has to happen on register operands. What you're saying is essentially the opposite of that, which you might call a [register memory architecture](https://en.wikipedia.org/wiki/Register_memory_architecture).